fix: support image source objects

jaulz opened the pull request:
This PR introduces support for Image source objects (related issue: https://github.com/necolas/react-native-web/issues/1019) which allows you to pass different headers or fetch data via POST method.

In the example I added a dynamic chart from Google Charts. Not directly related but to fill the row I added WebP and the mentioned redirect case in the issue.

![image](https://user-images.githubusercontent.com/5358638/68135005-43539480-ff23-11e9-9a0d-135b1066e0ad.png)

Please let me know if anything is missing.

necolas commented:
Thanks for the PR! The unit tests are currently failing but at a glance this direction looks good to me, and should allow onProgress to be implemented too. So I'll work with you to get this merged

jaulz commented:
@necolas great! 😊 I just fixed some minor issues and now all tests are running fine. Also, I added the `onProgress` prop but that triggers only once for the example. Maybe we can find a better example?

necolas commented on packages/react-native-web/src/exports/Image/ImageUriCache.js:
```diff
@@ -7,57 +7,80 @@
 
-  static has(uri: string) {
+  static createCacheId(source: ImageSource) {
+    return JSON.stringify(ImageLoader.resolveSource(source));
```
I'm a little concerned about using `JSON.stringify` here. We could use a Map instead

jaulz replied:
Yeah, I don't like it either but I could not find a smart way. The advantage of this approach is that you will always get the cached content for a combination of method, uri, headers and body. If we use a Map instead and use the source as a key then we somehow first need to get the reference to the source which leads to the same issue again. Or did you have something else in mind?

necolas replied:
A `Map` can use objects as keys

jaulz replied:
Yes, it allows object as keys but the reference is stored as a key and not the content of the object:
```ts
const map = new Map()
const key = { a: 1 }
map.set(key, true)

map.get(key) !== map.get({ a: 1 }) // true
```

necolas replied:
Yeah that's a problem given how image sources are defined. I guess we're stuck with stringifying for now

necolas commented on packages/react-native-web/src/exports/Image/index.js:
```diff
@@ -33,46 +32,28 @@ const STATUS_LOADING = 'LOADING';
-    uri = source;
-  } else if (source && typeof source.uri === 'string') {
-    uri = source.uri;
+const getCacheUrl = e => {
```
This looks a bit fragile. Why do we need to get the url from the native event?

jaulz replied:
If an image is loaded via XHR, the actual image `src` will be replaced by a blob:
https://github.com/necolas/react-native-web/blob/acd3b335518a362842ecfe006a796fcfd2c1a076/packages/react-native-web/src/modules/ImageLoader/index.js#L110

Since the `onload` event is triggered by the image itself, we can retrieve the new blob source in `Image` via the image `src`  in the event handler without passing additional parameters. Though, of course we could  also make this more explicit and pass the event and the new blob url as separate parameters in `onLoad`. What do you prefer?

necolas replied:
Can we store the blob when we get it, so we don't have to read it back from the native event?

jaulz replied:
Yep, the blob will be stored once the image was loaded:
```ts
ImageUriCache.add(source, getCacheUrl(e));
```

Afterwards it shouldn't be triggered again as long as the source does not change. Or do you mean something else?

necolas replied:
I'm wondering why we need to get the blob from the event object when we know what it is before that

jaulz replied:
I just reworked that part and now the blob is returned directly from the `ImageLoader`.

matthewwillian commented:
Hey is there any update with this PR?

jaulz commented:
@necolas I am just back from vacation and saw that there are conflicting files. Though, I don't see any changes in the affected files and am not sure why it says there are conflicts. Any idea what's wrong here?

necolas commented:
The conflicts are because the docs were rewritten. And now there is also a7ab961d95de892e4b41c96ea945aa344fd49e50

humaidk2 commented:
Hi, 
Any update on this?

necolas commented:
#2442 is a contemporary fix for this. Thanks
